File: js/src/fields.ts
/**
 * Fields (`docx_field`) per spec/algorithms.md §20.
 *
 * `insert_toc` adds, after the anchor, a paragraph holding the run-triple TOC
 * field (`TOC \o "1-{levels}" \h \z \u`). `insert_page_number` ensures the
 * section's footer exists (§15 machinery) and appends a `PAGE` field run-triple
 * to it. `update` sets `<w:updateFields w:val="true"/>` in word/settings.xml
 * (created on demand). Values materialize only at render — results never report
 * computed page/TOC numbers. The Python twin (`_fields.py`) is byte-parity.
 */
import { ToolError } from "./errors.js";
import {
  addRelationship,
  anchorInvalid,
  anchorNotFound,
  anchorStale,
  ensureContentOverride,
} from "./phase2common.js";
import type { DocHandle, Session } from "./session.js";
import { type ElementSlice, attrs, childElements, findElement, splice } from "./xmlscan.js";

const W_NS = "http://schemas.openxmlformats.org/wordprocessingml/2006/main";
const FOOTER_CT = "application/vnd.openxmlformats-officedocument.wordprocessingml.footer+xml";
const FOOTER_REL = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/footer";
const HEADER_CT = "application/vnd.openxmlformats-officedocument.wordprocessingml.header+xml";
const HEADER_REL = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/header";
const SETTINGS_PART = "word/settings.xml";
const SETTINGS_CT = "application/vnd.openxmlformats-officedocument.wordprocessingml.settings+xml";
const SETTINGS_REL = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/settings";

const FULL_ANCHOR_RE = /^P([1-9][0-9]*)#([0-9a-f]{4})$/;

// ---------------------------------------------------------------------------
// Field run-triples (§20)
// ---------------------------------------------------------------------------

/** A field run-triple: begin / instrText / separate / placeholder / end. */
function fieldRunTriple(instr: string, placeholder: string): string {
  const space = /^\s|\s$/.test(instr) ? ' xml:space="preserve"' : "";
  return (
    '<w:r><w:fldChar w:fldCharType="begin"/></w:r>' +
    `<w:r><w:instrText${space}>${instr}</w:instrText></w:r>` +
    '<w:r><w:fldChar w:fldCharType="separate"/></w:r>' +
    `<w:r><w:t>${placeholder}</w:t></w:r>` +
    '<w:r><w:fldChar w:fldCharType="end"/></w:r>'
  );
}

// ---------------------------------------------------------------------------
// docx_field
// ---------------------------------------------------------------------------

export interface DocxFieldArgs {
  doc_id: string;
  op: "insert_toc" | "insert_page_number" | "update";
  after?: string | undefined;
  levels?: number | undefined;
  scope?: "header" | "footer" | undefined;
  track_changes?: boolean | undefined;
  author?: string | undefined;
}

export interface DocxFieldResult {
  new_anchor?: string;
  updated?: number;
  note?: string;
}

export function docxField(session: Session, args: DocxFieldArgs): DocxFieldResult {
  const doc = session.get(args.doc_id);
  switch (args.op) {
    case "insert_toc":
      return fieldInsertToc(doc, args);
    case "insert_page_number":
      return fieldInsertPageNumber(doc, args);
    case "update":
      return fieldUpdate(doc);
    default:
      throw new ToolError("invalid_args", `docx_field: unknown op ${String(args.op)}.`, []);
  }
}

function paragraphEntries(doc: DocHandle) {
  return doc.anchorIndex().filter((e) => e.kind === "p");
}

function requireParagraph(doc: DocHandle, anchor: string) {
  const m = FULL_ANCHOR_RE.exec(anchor);
  if (!m) throw anchorInvalid(`Malformed anchor string: ${anchor}.`);
  const entry = paragraphEntries(doc)[Number(m[1]) - 1];
  if (entry === undefined) throw anchorNotFound(anchor);
  if (entry.anchor !== anchor) throw anchorStale(anchor);
  return entry;
}

// --- insert_toc --------------------------------------------------------------

function fieldInsertToc(doc: DocHandle, args: DocxFieldArgs): DocxFieldResult {
  if (args.after == null) throw anchorInvalid("op 'insert_toc' requires after.");
  const entry = requireParagraph(doc, args.after); // hash FIRST
  const levels = Math.max(1, Math.trunc(args.levels ?? 3));
  const instr = ` TOC \\o "1-${levels}" \\h \\z \\u `;
  const para = `<w:p>${fieldRunTriple(instr, "Right-click to update field.")}</w:p>`;

  const xml = doc.documentXml();
  const position = entry.block.end;
  doc.pkg.setPart(doc.documentPartName, splice(xml, position, position, para));
  doc.invalidate();

  const fresh = paragraphEntries(doc)[entry.ordinal]; // the inserted paragraph
  return {
    new_anchor: fresh ? fresh.anchor : args.after,
    note: `Inserted a TOC field (levels 1-${levels}) after ${args.after}.`,
  };
}

// --- insert_page_number ------------------------------------------------------

function fieldInsertPageNumber(doc: DocHandle, args: DocxFieldArgs): DocxFieldResult {
  const scope = args.scope ?? "footer";
  const partName = ensureScopePart(doc, scope);
  // Append a PAGE field paragraph to the header/footer part.
  const xml = doc.pkg.partText(partName);
  const rootTag = scope === "header" ? "w:hdr" : "w:ftr";
  const para = `<w:p>${fieldRunTriple(" PAGE ", "1")}</w:p>`;
  const close = xml.lastIndexOf(`</${rootTag}>`);
  let next: string;
  if (close >= 0) {
    next = splice(xml, close, close, para);
  } else {
    const selfClose = xml.lastIndexOf("/>");
    next = splice(xml, selfClose, selfClose + 2, `>${para}</${rootTag}>`);
  }
  doc.pkg.setPart(partName, next);
  return { note: `Inserted a PAGE field in the ${scope} (${partName}).` };
}

/**
 * Ensure the body section has a default header/footer of the given scope,
 * creating the part + content-type + rel + sectPr reference when absent.
 * Returns the part name to append the field paragraph into.
 */
function ensureScopePart(doc: DocHandle, scope: "header" | "footer"): string {
  const refName = scope === "header" ? "w:headerReference" : "w:footerReference";
  const xml = doc.documentXml();
  const bodySect = bodySectPr(xml);

  // If a default-variant reference already exists, reuse its part.
  if (bodySect && !bodySect.selfClosed) {
    for (const k of childElements(xml, bodySect.contentStart, bodySect.contentEnd)) {
      if (k.name === refName) {
        const a = attrs(xml, tagOf(k));
        if ((a["w:type"] ?? "default") === "default") {
          const rId = a["r:id"];
          if (rId) {
            const rel = doc.pkg.rels(doc.documentPartName).find((r) => r.id === rId);
            if (rel) {
              const target = rel.target.startsWith("/")
                ? rel.target.slice(1)
                : `word/${rel.target}`;
              if (doc.pkg.has(target)) return target;
            }
          }
        }
      }
    }
  }

  // Create a fresh part.
  const idx = nextHeaderFooterIndex(doc);
  const partName = `word/${scope}${idx}.xml`;
  const rootTag = scope === "header" ? "w:hdr" : "w:ftr";
  doc.pkg.setPart(
    partName,
    '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\n' + `<${rootTag} xmlns:w="${W_NS}"/>`,
  );
  ensureContentOverride(doc.pkg, partName, scope === "header" ? HEADER_CT : FOOTER_CT);
  const rId = addRelationship(
    doc.pkg,
    doc.documentPartName,
    scope === "header" ? HEADER_REL : FOOTER_REL,
    `${scope}${idx}.xml`,
  );

  // Splice the reference as the first child of the body sectPr.
  const refTag = `<${refName} w:type="default" r:id="${rId}"/>`;
  const next = insertBodyReference(doc.documentXml(), refTag);
  doc.pkg.setPart(doc.documentPartName, next);
  doc.invalidate();
  return partName;
}

function tagOf(el: ElementSlice) {
  return {
    kind: el.selfClosed ? ("empty" as const) : ("start" as const),
    name: el.name,
    start: el.start,
    end: el.startTagEnd,
    nameEnd: el.nameEnd,
  };
}

/** The trailing body-level w:sectPr, or null. */
function bodySectPr(xml: string): ElementSlice | null {
  const body = findElement(xml, "w:body");
  if (!body) return null;
  const kids = childElements(xml, body.contentStart, body.contentEnd);
  return kids.find((k) => k.name === "w:sectPr") ?? null;
}

function nextHeaderFooterIndex(doc: DocHandle): number {
  let max = 0;
  for (const name of doc.pkg.entryNames()) {
    const m = /^word\/(?:header|footer)([0-9]+)\.xml$/.exec(name);
    if (m) max = Math.max(max, Number(m[1]));
  }
  return max + 1;
}

/** Splice a reference as the first child of the trailing body sectPr. */
function insertBodyReference(xml: string, refTag: string): string {
  const sect = bodySectPr(xml);
  if (!sect) throw new ToolError("anchor_not_found", "Document has no body section.", []);
  if (sect.selfClosed) {
    return splice(xml, sect.start, sect.end, `<w:sectPr>${refTag}</w:sectPr>`);
  }
  return splice(xml, sect.contentStart, sect.contentStart, refTag);
}

// --- update ------------------------------------------------------------------

function fieldUpdate(doc: DocHandle): DocxFieldResult {
  ensureSettingsPart(doc);
  const xml = doc.pkg.partText(SETTINGS_PART);
  // Idempotent: if updateFields is already present, leave it.
  if (findElement(xml, "w:updateFields")) {
    return { updated: 1, note: "Fields already flagged for update." };
  }
  const entry = '<w:updateFields w:val="true"/>';
  const settings = findElement(xml, "w:settings");
  let next: string;
  if (settings && !settings.selfClosed) {
    // updateFields is the first child of w:settings.
    next = splice(xml, settings.contentStart, settings.contentStart, entry);
  } else if (settings && settings.selfClosed) {
    next = splice(
      xml,
      settings.start,
      settings.end,
      `<w:settings xmlns:w="${W_NS}">${entry}</w:settings>`,
    );
  } else {
    next = xml;
  }
  doc.pkg.setPart(SETTINGS_PART, next);
  return { updated: 1, note: "Flagged all fields for update on next render." };
}

function ensureSettingsPart(doc: DocHandle): void {
  if (doc.pkg.has(SETTINGS_PART)) return;
  doc.pkg.setPart(
    SETTINGS_PART,
    '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\n' + `<w:settings xmlns:w="${W_NS}"/>`,
  );
  ensureContentOverride(doc.pkg, SETTINGS_PART, SETTINGS_CT);
  addRelationship(doc.pkg, doc.documentPartName, SETTINGS_REL, "settings.xml");
}

File: src/docxengine/_fields.py
"""Fields (``docx_field``) — algorithms.md §20.

``insert_toc`` emits a paragraph holding the TOC run-triple field after an anchor
(never ``w:fldSimple``). ``insert_page_number`` ensures the body section's footer
(or header) exists and appends a ``PAGE`` field run-triple to it. ``update`` flips
``<w:updateFields w:val="true"/>`` in ``word/settings.xml`` (creating the part on
demand). All edits splice raw bytes per §3; values materialize only at render.
"""

from __future__ import annotations

import re

from . import _edits, _parts, _xml
from ._errors import ToolError
from ._opc import Package, resolve_rel_target
from ._session import Session

SETTINGS_PART = "word/settings.xml"
_SETTINGS_REL_TYPE = f"{_parts.REL_BASE}/settings"
_SETTINGS_CT = f"{_parts.CT_BASE}.settings+xml"
_HEADER_REL_TYPE = f"{_parts.REL_BASE}/header"
_FOOTER_REL_TYPE = f"{_parts.REL_BASE}/footer"
_HEADER_CT = f"{_parts.CT_BASE}.header+xml"
_FOOTER_CT = f"{_parts.CT_BASE}.footer+xml"
_W_NS_DECL = 'xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main"'

_WS_EDGE_RE = re.compile(f"^[{re.escape(_xml.WHITESPACE)}]|[{re.escape(_xml.WHITESPACE)}]$")


def _field_invalid(detail: str) -> ToolError:
    return ToolError("anchor_invalid", detail, ["Check the op arguments."])


# ---------------------------------------------------------------------------
# Field run-triples (§20)
# ---------------------------------------------------------------------------


def _field_run_triple(instr: str, placeholder: str) -> str:
    """A field run-triple: begin / instrText / separate / placeholder / end."""
    space = ' xml:space="preserve"' if _WS_EDGE_RE.search(instr) else ""
    return (
        '<w:r><w:fldChar w:fldCharType="begin"/></w:r>'
        f"<w:r><w:instrText{space}>{instr}</w:instrText></w:r>"
        '<w:r><w:fldChar w:fldCharType="separate"/></w:r>'
        f"<w:r><w:t>{placeholder}</w:t></w:r>"
        '<w:r><w:fldChar w:fldCharType="end"/></w:r>'
    )


# ---------------------------------------------------------------------------
# insert_toc (§20)
# ---------------------------------------------------------------------------


def _insert_toc(package: Package, after: str, levels: int) -> tuple[str, int]:
    main = package.main_document_part()
    entries = _edits.paragraph_entries(package)
    entry = _edits.require_paragraph(entries, after)
    safe_levels = max(1, int(levels))
    instr = f' TOC \\o "1-{safe_levels}" \\h \\z \\u '
    para = f"<w:p>{_field_run_triple(instr, 'Right-click to update field.')}</w:p>"
    data = package.part(main)
    position = entry.span.end
    package.set_part(main, _xml.splice(data, [(position, position, para.encode("utf-8"))]))
    fresh = _edits.paragraph_entries(package)
    new_anchor = fresh[entry.ordinal].anchor if entry.ordinal < len(fresh) else after
    return new_anchor, safe_levels


# ---------------------------------------------------------------------------
# insert_page_number (§20)
# ---------------------------------------------------------------------------


def _body_sectpr(data: bytes) -> _xml.Span | None:
    for child in _xml.iter_body_children(data):
        if child.name == "w:sectPr":
            return child
    return None


def _next_header_footer_index(package: Package) -> int:
    pattern = re.compile(r"^word/(?:header|footer)([0-9]+)\.xml$")
    max_n = 0
    for name in package.part_names:
        m = pattern.match(name)
        if m:
            max_n = max(max_n, int(m.group(1)))
    return max_n + 1


def _ensure_scope_part(package: Package, scope: str) -> str:
    """The body section's default ``scope`` part, creating it (+ wiring) when absent."""
    main = package.main_document_part()
    ref_name = "w:headerReference" if scope == "header" else "w:footerReference"
    data = package.part(main)
    body_sect = _body_sectpr(data)
    if body_sect is not None and not body_sect.empty:
        for ref in _xml.iter_elements(
            data, body_sect.inner_start, body_sect.inner_end, names=(ref_name,), max_depth=1
        ):
            attrs = _edits.start_tag_attrs(data, ref)
            if attrs.get("w:type", "default") == "default":
                rel_id = attrs.get("r:id")
                if rel_id:
                    for rel in package.rels(main):
                        if rel.rel_id == rel_id and not rel.is_external:
                            target = resolve_rel_target(main, rel.target)
                            if package.has_part(target):
                                return target
    # Create a fresh empty part and wire it.
    idx = _next_header_footer_index(package)
    part_name = f"word/{scope}{idx}.xml"
    root = "w:hdr" if scope == "header" else "w:ftr"
    body = (
        '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\r\n'
        f"<{root} {_W_NS_DECL}></{root}>"
    ).encode()
    package.set_part(part_name, body)
    _parts.ensure_content_type_override(
        package, part_name, _HEADER_CT if scope == "header" else _FOOTER_CT
    )
    rel_id = _parts.next_rel_id(package, main)
    _parts.add_relationship(
        package,
        main,
        rel_id,
        _HEADER_REL_TYPE if scope == "header" else _FOOTER_REL_TYPE,
        f"{scope}{idx}.xml",
    )
    # Splice the reference as the first child of the body sectPr.
    data = package.part(main)
    body_sect = _body_sectpr(data)
    if body_sect is None:
        raise ToolError("anchor_not_found", "Document has no body section.", [])
    ref_tag = f'<{ref_name} w:type="default" r:id="{rel_id}"/>'
    if body_sect.empty:
        new = _xml.splice(
            data, [(body_sect.start, body_sect.end, f"<w:sectPr>{ref_tag}</w:sectPr>".encode())]
        )
    else:
        new = _xml.splice(
            data, [(body_sect.inner_start, body_sect.inner_start, ref_tag.encode("utf-8"))]
        )
    package.set_part(main, new)
    return part_name


def _insert_page_number(package: Package, scope: str) -> str:
    part_name = _ensure_scope_part(package, scope)
    data = package.part(part_name)
    root = "w:hdr" if scope == "header" else "w:ftr"
    para = f"<w:p>{_field_run_triple(' PAGE ', '1')}</w:p>"
    close_tag = f"</{root}>".encode()
    close = data.rfind(close_tag)
    if close >= 0:
        new = _xml.splice(data, [(close, close, para.encode("utf-8"))])
    else:
        text = data.decode("utf-8")
        self_close = text.rfind("/>")
        new = (text[:self_close] + f">{para}</{root}>" + text[self_close + 2 :]).encode("utf-8")
    package.set_part(part_name, new)
    return part_name


# ---------------------------------------------------------------------------
# update (§20)
# ---------------------------------------------------------------------------


def _ensure_settings(package: Package) -> None:
    if package.has_part(SETTINGS_PART):
        return
    body = (
        '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\r\n'
        f"<w:settings {_W_NS_DECL}></w:settings>"
    ).encode()
    package.set_part(SETTINGS_PART, body)
    _parts.ensure_content_type_override(package, SETTINGS_PART, _SETTINGS_CT)
    main = package.main_document_part()
    _parts.add_relationship(
        package, main, _parts.next_rel_id(package, main), _SETTINGS_REL_TYPE, "settings.xml"
    )


def _update_fields(package: Package) -> tuple[int, str]:
    _ensure_settings(package)
    data = package.part(SETTINGS_PART)
    if _xml.find_element(data, "w:updateFields") is not None:
        return 1, "Fields already flagged for update."
    settings = _xml.find_element(data, "w:settings")
    if settings is None:
        return 1, "Flagged all fields for update on next render."
    fragment = '<w:updateFields w:val="true"/>'
    if settings.empty:
        replacement = f"<w:settings {_W_NS_DECL}>{fragment}</w:settings>".encode()
        new = _xml.splice(data, [(settings.start, settings.end, replacement)])
    else:
        new = _xml.splice(
            data, [(settings.inner_start, settings.inner_start, fragment.encode("utf-8"))]
        )
    package.set_part(SETTINGS_PART, new)
    return 1, "Flagged all fields for update on next render."


# ---------------------------------------------------------------------------
# docx_field
# ---------------------------------------------------------------------------


def docx_field(
    session: Session,
    *,
    doc_id: str,
    op: str,
    after: str | None = None,
    levels: int = 3,
    scope: str = "footer",
    track_changes: bool = False,
    author: str | None = None,
) -> dict[str, object]:
    """Insert TOC/page-number field codes or flag fields for update (§20)."""
    doc = session.get(doc_id)
    package = doc.package
    if op == "insert_toc":
        if after is None:
            raise _field_invalid("op 'insert_toc' requires after.")
        new_anchor, safe_levels = _insert_toc(package, after, levels)
        doc.mark_dirty()
        return {
            "new_anchor": new_anchor,
            "note": f"Inserted a TOC field (levels 1-{safe_levels}) after {after}.",
        }
    if op == "insert_page_number":
        part_name = _insert_page_number(package, scope)
        doc.mark_dirty()
        return {"note": f"Inserted a PAGE field in the {scope} ({part_name})."}
    if op == "update":
        updated, note = _update_fields(package)
        doc.mark_dirty()
        return {"updated": updated, "note": note}
    raise _field_invalid(f"Unknown field op: {op}.")
